services/src/main/java/org/keycloak/services/managers/DefaultBruteForceProtector.java

```diff
@@ -82,16 +85,58 @@ public ShutdownEvent() {
     protected class FailedLogin extends LoginEvent {
         protected final CountDownLatch latch = new CountDownLatch(1);
 
-        public FailedLogin(String realmId, String userId, String ip) {
-            super(realmId, userId, ip);
+        public FailedLogin(String realmId, String userId, ClientConnection clientConnection) {
+            super(realmId, userId, clientConnection);
         }
     }
 
     protected class SuccessfulLogin extends LoginEvent {
         protected final CountDownLatch latch = new CountDownLatch(1);
 
-        public SuccessfulLogin(String realmId, String userId, String ip) {
-            super(realmId, userId, ip);
+        public SuccessfulLogin(String realmId, String userId, ClientConnection clientConnection) {
+            super(realmId, userId, clientConnection);
+        }
+    }
+
+    protected static class AdaptedClientConnection implements ClientConnection {
```

Comment:
Could you please elaborate on why we need to copy values from client connection to instances of this class?

Is it related to using it from the background task and making sure the information is kept even though the request has ended?

Perhaps we could just have those fields in the `LoginEvent` itself rather than create a new class to just store them.

Reply:
I'm guessing they want the additional data for their SPI to email users. The rest of the PR seems to only read the remote address.

Reply:
@pedroigor I am not remember exactly why I have done this (it was two years ago :sweat_smile:) but by reading the code, I think it was I need to use the `EventBuilder` to send the event and the only constructor available takes the `ClientConnection` as 3rd args. See usage in `DefaultBruteForceProtector#failure`

```java
...
// Send event
new EventBuilder(realm, session, event.clientConnection)
...
```

Reply:
I see that. But still, not sure if we need a type wrapping the current `ClientConnection` but just pass the instance over.

Even though they are created from the worker thread (as part of the request lifecycle) I don't think we are going to have issues if we use it in the brute force background task.

Reply:
@paulbares By the way, I think your work is still valid and helps to improve auditing. Please, don't give up :)

Reply:
If change 
```java 
        protected LoginEvent(String realmId, String userId, ClientConnection clientConnection) {
            this.realmId = realmId;
            this.userId = userId;
            this.clientConnection = new AdaptedClientConnection(clientConnection);
        }
```
to this (using the same object):
```java 
        protected LoginEvent(String realmId, String userId, ClientConnection clientConnection) {
            this.realmId = realmId;
            this.userId = userId;
            this.clientConnection = clientConnection;
        }
```
All those tests are failing:
<img width="445" alt="image" src="https://github.com/keycloak/keycloak/assets/5783183/9791b486-b40c-49ed-a40a-630f94d2b45d">

The assertion errors are different each time.

Reply:
I see. So the fact it is being used in distinct threads is interfering. Please, ignore then.